lib/tbot/services/application/proxy_service.go

```diff
@@ -0,0 +1,377 @@
+	}
+	s.proxyAddr = proxyAddr
+
+	// Check if ALPN upgrade will be required to pass client certificate to the
+	// proxy.
+	s.alpnUpgradeRequired, err = s.alpnUpgradeCache.IsUpgradeRequired(
+		ctx, proxyAddr, s.connCfg.Insecure,
+	)
+	if err != nil {
+		return trace.Wrap(err, "testing if ALPN connection upgrade is required")
+	}
+
+	httpSrv := http.Server{
+		ReadTimeout:       apidefaults.DefaultIOTimeout,
+		ReadHeaderTimeout: defaults.ReadHeadersTimeout,
+		WriteTimeout:      apidefaults.DefaultIOTimeout,
+		IdleTimeout:       apidefaults.DefaultIdleTimeout,
+		BaseContext: func(net.Listener) context.Context {
+			// Use the main context which controls the service being stopped as
+			// the base context for all incoming requests.
+			return ctx
+		},
+	}
+	httpSrv.Handler = http.HandlerFunc(func(w http.ResponseWriter, req *http.Request) {
+		err := s.handleProxyRequest(w, req)
+		if err != nil {
+			trace.WriteError(w, err)
+			s.log.ErrorContext(
+				req.Context(), "Encountered an error while proxying request",
+				"error", err,
+			)
+			return
+		}
+	})
+	s.log.InfoContext(ctx, "Finished initializing")
+
+	var errCh = make(chan error, 1)
+	go func() {
+		s.log.DebugContext(ctx, "Starting proxy request handler goroutine")
+		errCh <- httpSrv.Serve(l)
+	}()
+	s.log.InfoContext(
+		ctx, "Listening for proxy connections",
+		"address", l.Addr().String(),
+	)
+	s.statusReporter.Report(readyz.Healthy)
+
+	select {
+	case <-ctx.Done():
+		return trace.Wrap(httpSrv.Close(), "closing http server")
+	case err := <-errCh:
+		s.statusReporter.ReportReason(readyz.Unhealthy, err.Error())
+		return trace.Wrap(err, "local proxy failed")
+	}
+}
+
+// String returns a human friendly representation of the service.
+func (s *ProxyService) String() string {
+	return cmp.Or(
+		s.cfg.Name,
+		fmt.Sprintf("%s:%s", ProxyServiceType, s.cfg.Listen),
+	)
+}
+
+// issueCert issues a role-impersonated app-routed X509 certificate
+func (s *ProxyService) issueCert(
+	ctx context.Context,
+	appName string,
+) (*tls.Certificate, types.Application, error) {
+	ctx, span := tracer.Start(ctx, "ProxyService/issueCert")
+	defer span.End()
+
+	// TODO(noah): At a later date, we should consider running a background
+	// goroutine that maintains a role-impersonated client. We should probably
+	// make this a generic helper since we have a few services that do this now.
+	// This will avoid the need to create and destroy a client for each new
+	// upstream.
+
+	// Right now we have to redetermine the route to app each time as the
+	// session ID may need to change. Once v17 hits, this will be automagically
+	// calculated by the auth server on cert generation, and we can fetch the
+	// routeToApp once.
+	effectiveLifetime := cmp.Or(s.cfg.CredentialLifetime, s.defaultCredentialLifetime)
+	identityOpts := []identity.GenerateOption{
+		identity.WithLifetime(effectiveLifetime.TTL, effectiveLifetime.RenewalInterval),
+		identity.WithLogger(s.log),
+	}
```

Comment:
Is this still relevant? (Maybe if we're backporting to v17?)

Reply:
Yeah if we're backporting we'll still need to do this. I'll come back and change to the new style for v18.